### DISCUSSION TITLE
System.InvalidOperationException: 'No service for type 'Quartz.IScheduler' has been registered.

### DISCUSSION THREAD

**Discussion by wineloy** [quartznet/quartznet#2188] [Q&A] [answered]
Hello everyone, I'm trying to add Quartz to a Windows Forms project with .NET 4.7.2 using the Microsoft.Extensions.DependencyInjection extension and Quartz.Extensions.DependencyInjection. However, in the Program.cs class:

```
var services = new ServiceCollection();

services.AddQuartz();

var serviceProvider = services.BuildServiceProvider();
var schedulerFactory = serviceProvider.GetRequiredService<IScheduler>();
```
However, on the line var schedulerFactory = serviceProvider.GetRequiredService<IScheduler>();, the following exception is thrown: System.InvalidOperationException: 'No service for type 'Quartz.IScheduler' has been registered.'

Does anyone know if there is an error? I haven't found anything on the web about this issue."

**Comment by jafin** [comment, marked as answer]
Is it a `ISchedulerFactory` that you need to be requesting?

Looking into the `ServiceCollectionExtensions.cs` it registers a ISchedulerFactory, but I can't see it registering IScheduler.

**Reply by lahma** [reply]
Yes, Quartz does not register `IScheduler` as getting it from the factory is an async operation. You need to get the `ISchedulerFactory` and then `await` the `GetScheduler()` for it.

For Windows Forms project I would presume that even the plain `SchedulerBuilder` API might suffice if DI is not needed.
